### DISCUSSION TITLE
Can I remove private attributes from vscode class signature?

### DISCUSSION THREAD

**Discussion by eelcovdw** [pydantic/pydantic#4563] [Question] [answered]
Using Python 3.9.12, Pydantic 1.10.2, VSCode 1.71.2, Pylance v2022.9.30

I have the following pydantic model:

```python
from pydantic import BaseModel, Extra
from typing import Optional


class Item(BaseModel):
    _private: Optional[str] = None
    uid: Optional[str] = None

    class Config:
        extra=Extra.forbid
        underscore_attrs_are_private = True

item = Item(_private="")
```

My editor has no issue with this code, and on checking the class signature of Item from pylance, it shows:

```
(class) Item(*, _private: str | None = None, uid: str | None = None)
```

However, when running this code, I get the following (expected) error:
```
ValidationError: 1 validation error for Item
_private
  extra fields not permitted (type=value_error.extra)
```

If I print the `__signature__` of the class, all looks as intended:

```
> print(Item.__signature__)

(*, uid: Optional[str] = None) -> None
```

Not a big problem when I only have one private attribute, but when I add multiple private attributes to a Model the signature the editor shows gets cluttered with many arguments I don't want to see:

```
(*, _private: str | None = None, _private2: Dict[str, int] | None = None, _private3: List[str] | None = None, _private4: int | None = None, _private5: bool | None = None, uid: str | None = None) -> None
```

Is there any way to make VSCode skip these private attrs in the generated Item signature? Or perhaps a workaround to define some non-field attributes somewhere else without them showing up?

**Comment by samuelcolvin** [comment, marked as answer]
This is because pyright/pylance/vscode uses [Dataclass Transforms](https://peps.python.org/pep-0681/) to interpret pydantic models as "like a dataclass" interms of signature.

Dataclass transforms only allow a minimal amount of customisation and AFAIK that doesn't include omitting private attributes from the signature.

Sorry to not be more help.

**Reply by eelcovdw** [reply]
Thank you @samuelcolvin !

If people are searching for the same issue, I found this (messy) workaround; define a PrivateAttr in the Config, add it via a new metaclass and add it under `TYPE_CHECKING` in the init for the type checker:

```python
from typing import TYPE_CHECKING, Any, Optional

from pydantic import BaseModel, Extra, PrivateAttr
from pydantic.fields import Field, FieldInfo
from pydantic.main import ModelMetaclass, __dataclass_transform__


@__dataclass_transform__(kw_only_default=True, field_descriptors=(Field, FieldInfo))
class PrivateModelMeta(ModelMetaclass):
    def __new__(mcs, name, bases, namespace, **kwargs) -> Any:
        cls = super().__new__(mcs, name, bases, namespace, **kwargs)

        config_from_namespace = namespace.get("Config")
        if config_from_namespace:
            additional_private_attrs = getattr(
                config_from_namespace, "additional_private_attrs", {}
            )
            cls.__private_attributes__ = {
                **cls.__private_attributes__,
                **additional_private_attrs,
            }
            cls.__slots__ = cls.__slots__ | additional_private_attrs.keys()
        return cls


class PrivateBaseModel(BaseModel, metaclass=PrivateModelMeta):
    def __init__(self, **data: dict) -> None:
        if TYPE_CHECKING:
            self._private: bool = True
        super().__init__(**data)

    class Config:
        extra = Extra.forbid
        additional_private_attrs = {"_private": PrivateAttr(True)}


class MyModel(PrivateBaseModel):
    uid: Optional[str] = None
```

Gives the same Model as:
```python
class MyModel(BaseModel):
    _private: bool = True
    uid: Optional[str] = None

    class Config:
        extra = Extra.forbid
        underscore_attrs_are_private = True
```

It's not the prettiest, haven't tested it extensively but it seems to work

**Reply by hassec** [reply]
This is now possible in pydantic, by using `PrivateAttr` [see docs](https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.PrivateAttr)
